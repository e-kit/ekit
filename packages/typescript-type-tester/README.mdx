---
name: TypeScript Type Tester
menu: '开发/测试/构建'
route: /ekit/typescript-type-tester
---

# TypeScript Type Tester

TypeScript 类型测试工具。

```
npm i -S @ekit/typescript-type-teste
```

## parser

通过检测示例代码内的 TypeScript 类型错误来进行对自定义工具类型或者 TypeScript 代码类型的单元测试，示例：

### xxOK.ts

```ts
// 符合静态类型检测，不应该抛出错误
const str: string = 'string';
const num: number = 1;
```

### xxFail.ts

```ts
// 不符合静态类型检测，应该抛出错误
const str: string = 1;
const num: number = 'string';
```

### xx.spec.ts 单测代码

```ts
import { parser } from '@ekit/typescript-type-tester';

describe('xx work ok', () => {
  const files = 'xxFail.ts xxOK.ts'.split(' ');
  const errors = parser(files, '__tests__');
  files.forEach((file) =>
    it(`check ${file} work ok`, () => {
      const current = errors[file];
      if (file.match(/OK/)) {
        expect(current).toHaveLength(0);
      } else if (file.match(/Fail/)) {
        expect(current.length).toBeGreaterThan(0);
      }
      expect(current).toMatchSnapshot();
    })
  );
});
```
