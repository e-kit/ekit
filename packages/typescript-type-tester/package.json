{
  "name": "@ekit/typescript-type-tester",
  "version": "0.0.4",
  "description": "test typescript type with jest and typescript",
  "keywords": [
    "typescript",
    "jest"
  ],
  "author": "yangqianjun",
  "homepage": "https://github.com/e-kit/ekittree/master/packages/typescript-type-tester#readme",
  "license": "MIT",
  "main": "lib/index.js",
  "directories": {
    "lib": "lib",
    "test": "__tests__"
  },
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/gogoyqj/ekit.git"
  },
  "scripts": {
    "test": "jest --all",
    "build": "tsc --project tsconfig.prod.json"
  },
  "bugs": {
    "url": "https://github.com/e-kit/ekitissues"
  },
  "peerDependencies": {
    "tslib": "*",
    "typescript": "^3.8.2"
  },
  "gitHead": "b395eb7c33153865c8fe74ff270b814d08805308"
}
