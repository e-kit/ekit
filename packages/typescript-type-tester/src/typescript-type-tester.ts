import * as path from 'path';
import * as ts from 'typescript';

/**
 * 解析一个 .ts 文件并返回错误信息
 * @param files 文件名
 * @param directory 绝对路径
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parser = (
  files: string[],
  directory: string,
  runtimeCompilerOptions: ts.CompilerOptions = {}
) => {
  /** 当前执行脚本的目录 */
  const cwd = process.cwd();
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const { compilerOptions } = require(path.join(cwd, 'tsconfig.json'));
  const fileNameMap: { [file: string]: any[] } = {};
  const errorMap: { [file: string]: any[] } = {};
  const program = ts.createProgram(
    files.map((file) => {
      const p = path.join(directory, file);
      fileNameMap[p] = errorMap[file] = [];
      return p;
    }),
    {
      ...compilerOptions,
      ...runtimeCompilerOptions,
      moduleResolution: ts.ModuleResolutionKind.NodeJs,
    }
  );
  const allDiagnostics = ts.getPreEmitDiagnostics(program);
  allDiagnostics.map((d) => {
    if (d.file) {
      if (d.file.fileName in fileNameMap) {
        const errors = fileNameMap[d.file.fileName];
        const { relatedInformation, code, category, messageText } = d;
        errors.push({
          fileName: path.relative(cwd, d.file.fileName),
          code,
          category,
          messageText: ts.flattenDiagnosticMessageText(messageText, '\n'),
        });
        if (Array.isArray(relatedInformation)) {
          relatedInformation.forEach((d) => {
            const { code, category, messageText, file } = d;
            // 方便测试
            console.log({
              fileName: path.relative(cwd, (file && file.fileName) || ''),
              code,
              category,
              messageText: ts.flattenDiagnosticMessageText(messageText, '\n'),
            });
          });
        }
      }
    }
  });
  return errorMap;
};
