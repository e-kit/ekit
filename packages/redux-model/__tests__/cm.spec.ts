/**
 * @file: CM 语法层面测试
 */
import { parser } from '@ekit/typescript-type-tester';

const files = 'testCMFail.ts testCMOK.ts'.split(' ');

describe('utils/useModel work ok', () => {
  const errors = parser(files, '__tests__/cmSamples');
  files.forEach((file) =>
    it(`check cmSamples/${file} work ok`, () => {
      const current = errors[file];
      if (file.match(/OK/)) {
        expect(current).toHaveLength(0);
      } else if (file.match(/Fail/)) {
        expect(current.length).toBeGreaterThan(0);
      }
      expect(current).toMatchSnapshot();
    })
  );
});
