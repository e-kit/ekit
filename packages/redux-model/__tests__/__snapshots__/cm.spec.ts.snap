// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/useModel work ok check cmSamples/testCMFail.ts work ok 1`] = `
Array [
  Object {
    "category": 1,
    "code": 2322,
    "fileName": "__tests__/cmSamples/testCMFail.ts",
    "messageText": "Type 'number' is not assignable to type 'string'.",
  },
  Object {
    "category": 1,
    "code": 2769,
    "fileName": "__tests__/cmSamples/testCMFail.ts",
    "messageText": "No overload matches this call.
  Overload 1 of 4, '(effect: <P extends Tction<number>>(payload: P[\\"payload\\"]) => P, args: number): Promise<Tction<number>>', gave the following error.
    Argument of type 'string' is not assignable to parameter of type 'number'.
  Overload 2 of 4, '(effect: <P extends Tction<number>>(payload: P[\\"payload\\"]) => P, payload: number): Promise<Tction<number>>', gave the following error.
    Argument of type 'string' is not assignable to parameter of type 'number'.",
  },
  Object {
    "category": 1,
    "code": 2345,
    "fileName": "__tests__/cmSamples/testCMFail.ts",
    "messageText": "Argument of type '2' is not assignable to parameter of type 'string'.",
  },
  Object {
    "category": 1,
    "code": 2322,
    "fileName": "__tests__/cmSamples/testCMFail.ts",
    "messageText": "Type '2' is not assignable to type 'string'.",
  },
  Object {
    "category": 1,
    "code": 2345,
    "fileName": "__tests__/cmSamples/testCMFail.ts",
    "messageText": "Argument of type '\\"2\\"' is not assignable to parameter of type 'number'.",
  },
  Object {
    "category": 1,
    "code": 2322,
    "fileName": "__tests__/cmSamples/testCMFail.ts",
    "messageText": "Type '\\"2\\"' is not assignable to type 'number'.",
  },
]
`;

exports[`utils/useModel work ok check cmSamples/testCMOK.ts work ok 1`] = `Array []`;
